### DISCUSSION TITLE
Consider if we should use Pydantic to type-check user's arguments to functions

### DISCUSSION THREAD

**Discussion by rht** [mesa/mesa#2166] [Ideas]
Pydantic has a validation decorator used to decorate functions so that the types from argument and return value annotations are enforced: https://docs.pydantic.dev/latest/concepts/validation_decorator/. The V2 has performance improvements that should be considered.

cc: @adamamer20 since mesa-frames has extensive type annotations.

**Comment by adamamer20** [comment]
I think that's a great idea, I also thought about it for mesa-frames but when i tried with [typeguard](https://pypi.org/project/typeguard/), there was a significant performance overhead. After we have some examples, we can test out pydantic to see whether the performance hit is noticeable or not.

**Reply by rht** [reply]
How did you test the performance? Is it by generating the readme_plot output? I want to try Pydantic on mesa-frames.

**Reply by adamamer20** [reply]
Yeah but it was quite a while ago before the new PRs, and it was more anecdotally than a rigorous test (I don't have any measurement for it). We can definitely try, look also at [beartype](https://beartype.readthedocs.io/en/latest/eli5/).

> Unlike [comparable runtime type-checkers](https://beartype.readthedocs.io/en/latest/moar/#moar-runtime) (e.g., [pydantic](https://pydantic-docs.helpmanual.io/), [typeguard](https://github.com/agronholm/typeguard)), beartype decorates callables with dynamically generated wrappers efficiently type-checking each parameter passed to and value returned from those callables in constant time. Since “performance by default” is our first-class concern, generated wrappers are guaranteed to:
> - Exhibit [O(1) non-amortized worst-case time complexity with negligible constant factors](https://beartype.readthedocs.io/en/latest/math/#math-time).
> - Be either more efficient (in the common case) or exactly as efficient minus the cost of an additional stack frame (in the worst case) as equivalent type-checking implemented by hand, which no one should ever do.
